google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatchingPublisher.java

```diff
@@ -0,0 +1,202 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.internal.wire;
+
+import static com.google.cloud.pubsublite.internal.CheckedApiPreconditions.checkState;
+import static com.google.cloud.pubsublite.internal.ExtractStatus.toCanonical;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.ApiService;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.cloud.pubsublite.Message;
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.PublishMetadata;
+import com.google.cloud.pubsublite.internal.*;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.stream.LongStream;
+
+public class PartitionCountWatchingPublisher extends ProxyService
+    implements Publisher<PublishMetadata> {
+  private static final GoogleLogger log = GoogleLogger.forEnclosingClass();
+  private final PartitionPublisherFactory publisherFactory;
+  private final RoutingPolicy.Factory policyFactory;
+
+  private static class PartitionsWithRouting {
+    public final ImmutableMap<Partition, Publisher<PublishMetadata>> publishers;
+    private final RoutingPolicy routingPolicy;
+
+    private PartitionsWithRouting(
+        ImmutableMap<Partition, Publisher<PublishMetadata>> publishers,
+        RoutingPolicy routingPolicy) {
+      this.publishers = publishers;
+      this.routingPolicy = routingPolicy;
+    }
+
+    public ApiFuture<PublishMetadata> publish(Message message) throws CheckedApiException {
+      try {
+        Partition routedPartition =
+            message.key().isEmpty()
+                ? routingPolicy.routeWithoutKey()
+                : routingPolicy.route(message.key());
+        checkState(
+            publishers.containsKey(routedPartition),
+            String.format(
+                "Routed to partition %s for which there is no publisher available.",
+                routedPartition));
+        return publishers.get(routedPartition).publish(message);
+      } catch (Throwable t) {
+        throw toCanonical(t);
+      }
+    }
+
+    public void flush() throws IOException {
+      for (Publisher<PublishMetadata> publisher : publishers.values()) {
+        publisher.flush();
+      }
+    }
+
+    public void stop() {
+      publishers.values().forEach(ApiService::stopAsync);
+      publishers.values().forEach(ApiService::awaitTerminated);
+    }
+  }
```

Comment:
So I spent a while trying to do that. The awkward bit is that RoutingPublisher is a ProxyApiService which contains all of its partition level publishers.

This makes it challenging to install an new routing publisher when the number of partitions changes, since it's not currently possible to add started services to a ProxyApiService, and it's not possible to remove services either.